plugins/zify/zify.ml

```diff
@@ -0,0 +1,1121 @@
+        Injterm
+          (EConstr.mkApp
+             ( force mkapp
+             , [| u.source1
+                ; u.source2
+                ; u.target
+                ; f
+                ; u.inj1_t
+                ; u.inj2_t
+                ; unop
+                ; e1 |] ))
+  else
+    let e1 = inj_term_of_texpr evd e1 in
+    Injterm
+      (EConstr.mkApp
+         ( force mkapp
+         , [|u.source1; u.source2; u.target; f; u.inj1_t; u.inj2_t; unop; e1|]
+         ))
+
+type typed_constr = {constr: EConstr.t; typ: EConstr.t}
+
+type op =
+  | Unop of
+      { unop: EConstr.t
+      ; (* unop : typ unop_arg -> unop_typ *)
+        unop_typ: EConstr.t
+      ; unop_arg: typed_constr }
+  | Binop of
+      { binop: EConstr.t
+      ; (* binop : typ binop_arg1 -> typ binop_arg2 -> binop_typ *)
+        binop_typ: EConstr.t
+      ; binop_arg1: typed_constr
+      ; binop_arg2: typed_constr }
+
+
+let rec trans_expr env evd e =
+  (* Get the injection *)
+  let {decl= i; deriv= inj} = InjTable.get evd e.typ in
+  let e = e.constr in
+  if EConstr.isConstruct evd e then Constant (inj, e) (* Evaluate later *)
+  else
+    try
+      (* The term [e] might be a registered constant *)
+      let {decl= c} = CstOp.get evd e in
+      Injterm
+        (EConstr.mkApp (force mkapp0, [|inj.source; inj.target; e; i; c|]))
+    with Not_found -> (
+      (* Let decompose the term *)
+      match EConstr.kind evd e with
+      | App (t, a) -> (
+        try
+          match Array.length a with
+          | 1 ->
+              let {decl= unop; deriv= u} = UnOp.get evd t in
+              let a' = trans_expr env evd {constr= a.(0); typ= u.source1} in
+              if is_constant a' && EConstr.isConstruct evd t then
+                Constant (inj, e)
+              else mkapp_id evd i inj (unop, u) t a'
+          | 2 ->
+              let {decl= bop; deriv= b} = BinOp.get evd t in
+              let a0 =
+                trans_expr env evd {constr= a.(0); typ= b.EBinOp.source1}
+              in
+              let a1 =
+                trans_expr env evd {constr= a.(1); typ= b.EBinOp.source2}
+              in
+              if is_constant a0 && is_constant a1 && EConstr.isConstruct evd t
+              then Constant (inj, e)
+              else mkapp2_id evd i inj t bop b a0 a1
+          | _ -> Var (inj, e)
+        with Not_found -> Var (inj, e) )
+      | _ -> Var (inj, e) )
+
+let trans_expr env evd e =
+  try trans_expr env evd e with Not_found ->
+    raise
+      (CErrors.user_err
+         ( Pp.str "Missing injection for type "
+         ++ Printer.pr_leconstr_env env evd e.typ ))
+
+let get_rel env evd e =
+  let is_prop env evd p =
+    let ty = get_type_of env evd p in
+    Constr.is_Prop (unsafe_to_constr ty)
+  in
+  let is_arrow a p1 p2 =
+    (* cheap *)
+    match a.Context.binder_name with
+    | Names.Anonymous -> is_prop env evd p1 && is_prop env evd p2
+    | _ ->
+        is_prop env evd p1 && is_prop env evd p2
```

Comment:
Isn't the env wrong for p2?
I guess this should be more like
~~~ocaml
let is_arrow a p1 p2 = 
  is_prop env evd p1 && is_prop (push_rel a env) evd p2 && (a.binder_name = Anonymous || noccurn evd 1 p2)
~~~

Reply:
Thanks. Done.